### ISSUE TITLE
Order dependence in a where clause

### ISSUE THREAD

**Issue by alanz** [ucsd-progsys/liquidhaskell#704]
I am working through the tutorial at https://ucsd-progsys.github.io/liquidhaskell-tutorial/04-poly.html using GHC 7.10.3  and z3 4.4.1

I find that the checker either fails or succeeds based on the order of definitions in a where clause

```haskell
{-@ LIQUID "--short-names"    @-}
{-@ LIQUID "--no-termination" @-}

import Prelude      hiding (head, abs, length)
import Data.List    (foldl')
import Data.Vector  hiding (head, foldl')

{-@ abs :: Int -> Nat @-}
abs           :: Int -> Int
abs n
  | 0 < n     = n
  | otherwise = 0 - n

{-@ type Btwn Lo Hi  = {v:Int      | Lo <= v && v < Hi} @-}
{-@ loop :: lo:Nat -> hi:{Nat|lo <= hi} -> a -> (Btwn lo hi -> a -> a) -> a @-}
loop :: Int -> Int -> a -> (Int -> a -> a) -> a
loop lo hi base f =  go base lo
  where
    go acc i
      | i < hi    = go (f i acc) (i + 1)
      | otherwise = acc


-- >>> absoluteSum' (fromList [1, -2, 3])
-- 6
{-@ absoluteSum' :: Vector Int -> Nat @-}
absoluteSum' :: Vector Int -> Int
absoluteSum' vec = loop 0 n 0 body
  where
    n            = length vec
    body i acc   = acc + (abs ( vec ! i))
```

In the last two lines, switching the order of definition for `n` and `body` changes the checker from passing all, to reporting 

```
 poly-bug.hs:36:33-39: Error: Liquid Type Mismatch
 
 36 |     body i acc   = acc + (abs ( vec ! i))
                                      ^^^^^^^
 
   Inferred type
     VV : {VV : GHC.Types.Int | VV == i}
  
   not a subtype of Required type
     VV : {VV : GHC.Types.Int | VV >= 0
                                && VV < vlen vec}
  
   In Context
     vec := {vec : (Data.Vector.Vector GHC.Types.Int) | 0 <= vlen vec}

     i := {i : GHC.Types.Int | i >= 0}
```

And it does not infer the annotation for `loop`

**Comment by ranjitjhala** [comment]
Hi @alanz, 

thanks for pointing this out! 

TL;DR: run with `--scrape-used-imports` or `--eliminate` to 
get it to work in both cases.

This is a subtle issue that needs more attention in the tutorial.
The reason for this difference is that with the "good" order, 
we can synthesize the refinement:

```
{-@ body :: i:{Nat | i < n} -> Nat -> Nat @-}
```

but with the "bad" order, `n` is not in scope at `body` and so 
we cannot get the above refinement.

Now, you may wonder, why not just synthesize the refinement:

```
{-@ body :: i:{Nat | i < vLength vec} -> Nat -> Nat @-}
```

where `i` is upper bounded by the size of `vec` which *is*
in scope in both cases? The reason is that the inference 
mechanism works with *qualifiers* which are 
"predicate fragments" and requires the fragment 
`(a < vLength b)` which is nowhere in this file. 

However, when you run with:

1. `--scrape-used-imports` , it finds this fragment from 
    the specification for `V.!`, or with

2. `--eliminate`, it runs a new experimental inference
    algorithm that doesn't require the qualifiers.

One of the above options will be made the default soon...

Thanks again for pointing this out! Did the above help?

Best!

Ranjit.

**Comment by alanz** [comment]
Does liquidhaskell process the definitions in lexical order? I know the GHC RenamedSource re-orders them in dependency order, which might be something worth considering.

**Comment by alanz** [comment]
Thinking about the tutorial, the `vectorSum'` example used the "good" order, but the one to be completed by the user (`absoluteSum'`) has the "bad" order, which is setting up the user to fail.

**Comment by gridaphobe** [comment]
We process the Core, by which point I believe `n` and `body` will no longer be in the same binder group. Furthermore, neither depends on the other, so GHC is free to order them however it feels. So I suspect if we look at the Core, we'll see that `body` is defined before `n`, which sadly means that `n` will not be in scope from LH's perspective.

**Comment by ranjitjhala** [comment]
Yes, LH operates at the lexical order, in particular, at
the level of the GHC core (and then processes the bindings
in that order). I'm guessing in one case, the core looks like:

```
 let n    = ...
     body = ...
 in
   ...
```

(so `n` is in scope at the definition of `body`), but in
the other case its:


```
 let body = ...
     n    = ...
 in
   ...
```

where `n` is not.

Note that in this case `n` doesn't actually appear in `body`
i.e. the only dependency is at the refinement level (which GHC
is oblivious to. :))

**Comment by ranjitjhala** [comment]
Ugh, you're right! I should fix that ASAP! Thanks!!

**Comment by alanz** [comment]
Btw, both the suggested fixes (`--scrape-used-options`, `--elimate`) work. Is there any reason to prefer the one over the other?

**Comment by ranjitjhala** [comment]
Right now, the behavior of `--scrape-used-imports`
is far better understood & predictable.

The second option, `--eliminate` is still under
development, and I'd like the dust to settle a
bit before recommending it.
